### PR TITLE
DOC: reorganize developer docs, use scikit-image as a base for change

### PR DESCRIPTION
**Pull request by mattip**
Following discussion in the weekly dev meetings about missing merge policy and general developer guidelines, I reworked the developer document layout and added content adapted from [`scikit-image`](https://github.com/scikit-image/scikit-image/blob/64103e6c90917fcfdef8343fd7dd4df32c910446/CONTRIBUTING.txt)

### CODE REVIEW

**Comment by mattip** [comment]
still a WIP, more editing needed

**Comment by rgommers** [comment]
Thanks Matti, this looks much more readable than gitwash. I like this and the scikit-image docs it's based on.

Note that this is the "contributing" guide, which is conceptually different from a developer guide with
detailed guidelines and policies on, e.g., how to merge PRs (when squashing etc.), how to make a release, a deprecation policy, etc. I suggest to keep those two things separate. Here are two examples I can think of:
- [AstroPy docs](http://docs.astropy.org/en/stable/index.html#developer-documentation) (contrib + dev together) 
- [SciPy docs](http://scipy.github.io/devdocs/index.html#developer-s-guide) (separate, "Contributing to SciPy" and "SciPy developer guide")

**Comment by rgommers** [comment]
Another alternative is to do the "developer guide" as a series of informational NEPs, e.g.:
- backwards compatibility ([NEP 23](http://www.numpy.org/neps/nep-0023-backwards-compatibility.html))
- releasing new NumPy versions (convert [HOWTO_RELEASE](https://github.com/numpy/numpy/blob/master/doc/HOWTO_RELEASE.rst.txt) to a NEP)
- coding guidelines (convert [C_STYLE_GUIDE](https://github.com/numpy/numpy/blob/master/doc/C_STYLE_GUIDE.rst.txt) to a NEP)
- etc.

Maybe that's the best option.

File: doc/source/dev/index.rst
Comment (mhvk) [inline doc/source/dev/index.rst]: Is this up to date? Wouldn't one use `pip install` these days? (That does work for me)

**Review by mhvk — COMMENTED** [review]
Great to see this update!  Some in-line comments.

File: doc/source/dev/index.rst
Comment (rgommers) [reply doc/source/dev/index.rst]: it is possible, nothing to stop you hitting the green button

Comment (rgommers) [reply doc/source/dev/index.rst]: anyway, I don't think this belongs here at all. contributors cannot merge PRs, so it's just a distraction here. should be in the separate maintainer guide or an informational NEP

File: doc/source/dev/index.rst
Comment (rgommers) [reply doc/source/dev/index.rst]: I'd prefer to keep this short, there are way too many popular editors to give instructions and "trailing whitespace" is easy to search for

File: doc/source/dev/index.rst
Comment (eric-wieser) [reply doc/source/dev/index.rst]: +1. At least in the source code, I've been using `gh-NNN`, because I've been bitten multiple times by `#NNN` referring to a trac issue number.

File: doc/source/dev/index.rst
Comment (eric-wieser) [inline doc/source/dev/index.rst]: nit: `upstream/master`

File: doc/source/dev/index.rst
Comment (eric-wieser) [reply doc/source/dev/index.rst]: +1, this belongs somewhere else.

File: doc/source/dev/index.rst
Comment (eric-wieser) [inline doc/source/dev/index.rst]: I've seen / been responsible for a handful of merges with just one reviewer. Are we making this against policy from now on, or sticking with a "use judgement" based approach where 2 is the norm, but trivial fixes are fine with just one?

Comment (rgommers) [reply doc/source/dev/index.rst]: No, this is just a copy-paste error - one reviewer is perfectly fine

Comment (rgommers) [reply doc/source/dev/index.rst]: scikit-image has always had a different policy

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: Agreed it shouldn't be here; maybe just change to "Set up your editor to follow PEP-8 (remove trailing white space, no tabs, etc.)" so that people google the right thing.

For the record, I found out about how nice emacs could be through the astropy link. Don't necessarily count on people understanding their editor can do it for them!

**Comment by mattip** [comment]
> Note that this is the "contributing" guide, which is conceptually different from a developer guide with detailed guidelines and policies on, e.g., how to merge PRs (when squashing etc.), how to make a release, a deprecation policy, etc. I suggest to keep those two things separate.

I think  contributors need to know what to expect: how to squash their commits,  PR merge policy is, and how to deprecate a feature. 

I agree making a release is advanced, that is why I only linked to it at the bottom of the document.

File: doc/source/dev/gitwash/git_intro.rst
Comment (mattip) [reply doc/source/dev/gitwash/git_intro.rst]: hmm. Maybe we shouldn't have this at all. No matter what we do it will get dated.

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: I think they are automatically subscribed to PRs they submit. Removing.

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: deleting

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removing

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: Linking into the C style guide

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removing, the link has installation instructions

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: fixing

**Comment by mattip** [comment]
I went through all the comments. The remaining one is a decision about scope: how to split our documentation into newish contributor/core developer. Could the reviewers make another pass through and "resolve conversations" where appropriate?

**Comment by mattip** [comment]
[Here](https://6843-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/dev/index.html) is the main page, at the bottom are the "in depth" links, including a rearranged [gitwash](https://6843-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/dev/gitwash/index.html)

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: either leave out or also add conda. there are too many ways to install, and we recommend anaconda for beginning users

Comment (rgommers) [reply doc/source/dev/index.rst]: also `-r` is not needed

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: this needs removing. if we keep the latex mention at all, it should just list package names or link to another page with more details. too many ways to do this, and too much of a hurdle

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: mention of the first two warnings are useful, this one is not I think - plots get included directly, there should not be external images normally

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: this can probably be removed, there was already a link to the NEP higher up

**Review by rgommers — DISMISSED** [review]
This reads well. The structure is good I think, my comments are mainly on textual things. The exception is the end - the discussions on LaTeX and deprecations can be moved out to keep it short and to the point for beginning contributors.

**Comment by rgommers** [comment]
> Here is the main page

There's a stray ToC at the bottom of the page

File: doc/source/dev/gitwash/git_intro.rst
Comment (mhvk) [reply doc/source/dev/gitwash/git_intro.rst]: I agree, so let's just go with `linux Use your favourite package manager to install "git"`

Comment (mhvk) [reply doc/source/dev/gitwash/git_intro.rst]: Arguably, the whole bullet-point list can be removed - or is that what you meant also?

Comment (mhvk) [reply doc/source/dev/gitwash/git_intro.rst]: Possibly the git resources also have installation instructions?

File: doc/source/dev/gitwash/index.rst
Comment (mhvk) [inline doc/source/dev/gitwash/index.rst]: What is this? I cannot find it in the current doc tree.  If related to python2, it should be removed.

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: I'd remove the "after running..." - generally that is not needed and you already link to a presumably more in-depth discussion.

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: Since we refer to the numpy C docs right below, should we just remove this here?

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: Agree, I don't think I've ever done anything but `make html` myself...

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: Is scipy needed for numpy docs?

Comment (rgommers) [reply doc/source/dev/index.rst]: just for the intersphinx links, there are some references

File: doc/source/dev/gitwash/git_intro.rst
Comment (mattip) [reply doc/source/dev/gitwash/git_intro.rst]: removed the bullet-point list in favor of an external link

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: adopted

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removed the `git clean` suggestion, but I left in the "Be sure to document ..."

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: fixed

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: nice, I didn't know [`gh-` works](https://help.github.com/en/articles/autolinked-references-and-urls#issues-and-pull-requests).  Changed

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok. I would prefer a week, but will go with consensus

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removing

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: related to #13336 and the recent experience with Outreachy interns. If that PR is accepted, the need to install numpy will self-document. Changed to explain in words, without `make dist`

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: ok

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: there is a docstring in [`lib/function_base.vectorize`](www.numpy.org/devdocs/reference/generated/numpy.vectorize.html#numpy.vectorize) at the bottom that requires scipy

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removed

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removing

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: removing

File: doc/source/dev/gitwash/index.rst
Comment (mattip) [reply doc/source/dev/gitwash/index.rst]: It is related to difference in git between `ref1..ref2` to `ref1...ref2`

http://www.numpy.org/devdocs/dev/gitwash/dot2_dot3.html

**Comment by mattip** [comment]
> There's a stray ToC at the bottom of the page

Since this is the index page, that is meant to be the complete developer documentation, the text above is a shortened summary.

**Comment by mattip** [comment]
I tried to address the new comments. Could the reviewers please make another pass through and "resolve conversations" where appropriate.

File: doc/source/dev/index.rst
Comment (rgommers) [reply doc/source/dev/index.rst]: you missed one instance of `transform-speedups` further rdown

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: members --> member

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: typo: in in

Comment (mattip) [reply doc/source/dev/index.rst]: fixed

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: fixed, reflowed

File: doc/source/dev/index.rst
Comment (rgommers) [reply doc/source/dev/index.rst]: on second thought, just removing install command is better. just say something like "requires Sphinx, Matplotlib and SciPy"

File: doc/source/dev/index.rst
Comment (mattip) [reply doc/source/dev/index.rst]: fixed, sorry

File: doc/source/dev/index.rst
Comment (rgommers) [inline doc/source/dev/index.rst]: this one should be `linspace-speedups`

File: doc/source/dev/index.rst
Comment (mhvk) [reply doc/source/dev/index.rst]: If that is the only reason for the `scipy` dependence, we should remove it! (Fine to do in another PR.)

File: doc/source/dev/gitwash/index.rst
Comment (mhvk) [reply doc/source/dev/gitwash/index.rst]: OK, thanks, learn something new every day!

File: doc/source/dev/index.rst
Comment (rgommers) [reply doc/source/dev/index.rst]: Ah yes, agree. I thought it was just an intersphinx link; this example can be done without scipy

**Comment by mhvk** [comment]
Am observing so no real change to look at this - but it looked in good shape and am happy to let @rgommers do the approval/merging.

**Comment by rgommers** [comment]
Looks great, in it goes. Thanks @mattip! And thanks all reviewers as well.

The SciPy dependency comment can be dealt with separately.

**Comment by rgommers** [comment]
There's also still [this comment](https://github.com/numpy/numpy/pull/13367#issuecomment-485072004) on what to do about developer docs. I'm tending towards a NEP per topic. Would match what Python does - we unfortunately can't match numbering, so no NEP 7 to match PEP 7:)

**Comment by stefanv** [comment]
I think you do want NEPs to address issues such as number of reviewers before merging.  Docs currently state: "A PR must be approved by at least one core team member before merging. Approval means the core team member has carefully reviewed the changes, and the PR is ready for merging."

This means that, if a core contributor reviews incorrectly, that change can go in.  Could be OK, but having a different person click the merge button from the one who reviewed is often a good idea.  Two eyes and all that...

File: doc/source/dev/index.rst
Comment (stefanv) [inline doc/source/dev/index.rst]: Typo: unmatched backtick

File: doc/source/dev/index.rst
Comment (stefanv) [inline doc/source/dev/index.rst]: This is RST, not md!

**Comment by rgommers** [comment]
> This means that, if a core contributor reviews incorrectly, that change can go in. Could be OK, but having a different person click the merge button from the one who reviewed is often a good idea. Two eyes and all that...

I believe that this may catch some extra things, but very few things that aren't fixable afterwards. It won't really help with catching subtle changes in behavior for example. Unless there's two very in-depth reviews, and we don't even have the bandwidth to keep up with the current load so doubling it is not the best idea.

**Comment by stefanv** [comment]
On Wed, 24 Apr 2019 01:41:55 -0700, Ralf Gommers wrote:
> I believe that this may catch some extra things, but very few things that aren't fixable afterwards. It won't really help with catching subtle changes in behavior for example. Unless there's two very in-depth reviews, and we don't even have the bandwidth to keep up with the current load so doubling it is not the best idea.

My point was more (in support of your recommendation) that this is a
policy decision, and as such should be debated as / published as a NEP.

**Pull request merged**
